### PR TITLE
Rename trainer arg `tokenizer` to `processing_class`

### LINKED ISSUE #2161: AttributeError: property 'tokenizer' of 'DPOTrainer' object has no setter
### System Info

- Platform: Linux-5.15.0-1048-aws-x86_64-with-glibc2.31
- Python version: 3.11.9
- PyTorch version: 2.4.1
- CUDA device: NVIDIA H100 80GB HBM3
- Transformers version: 4.46.0.dev0
- Accelerate version: 0.34.2
- Accelerate config: not found
- Datasets version: 3.0.0
- HF Hub version: 0.24.7
- TRL version: 0.12.0.dev0+07cebf3
- bitsandbytes version: 0.41.1
- DeepSpeed version: 0.15.1
- Diffusers version: 0.30.3
- Liger-Kernel version: 0.3.0
- LLM-Blender version: 0.0.2
- OpenAI version: 1.46.0
- PEFT version: 0.12.0

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoTokenizer, AutoModelForCausalLM
from datasets import load_dataset
import tempfile
from trl import DPOTrainer, DPOConfig

model_id = "HuggingFaceM4/tiny-random-LlamaForCausalLM"
tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(model_id)

with tempfile.TemporaryDirectory() as tmp_dir:
    training_args = DPOConfig(output_dir=tmp_dir)
    dummy_dataset = load_dataset("trl-internal-testing/zen", "standard_preference")
    trainer = DPOTrainer(model=model, args=training_args, tokenizer=tokenizer, train_dataset=dummy_dataset["train"])
```

```
[2024-10-03 09:03:00,224] [INFO] [real_accelerator.py:203:get_accelerator] Setting ds_accelerator to cuda (auto detect)
You are using the default legacy behaviour of the <class 'transformers.models.llama.tokenization_llama_fast.LlamaTokenizerFast'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565 - if you loaded a llama tokenizer from a GGUF file you can ignore this message.
/fsx/qgallouedec/transformers/src/transformers/generation/configuration_utils.py:579: UserWarning: `pad_token_id` should be positive but got -1. This will cause errors when batch generating, if there is padding. Please set `pad_token_id` explicitly as `model.generation_config.pad_token_id=PAD_TOKEN_ID` to avoid errors in generation
  warnings.warn(
Traceback (most recent call last):
  File "/fsx/qgallouedec/transformers/../trl/dfg.py", line 13, in <module>
    trainer = DPOTrainer(model=model, args=training_args, tokenizer=tokenizer, train_dataset=dummy_dataset["train"])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/fsx/qgallouedec/miniconda3/envs/trl/lib/python3.11/site-packages/huggingface_hub/utils/_deprecation.py", line 101, in inner_f
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/fsx/qgallouedec/trl/trl/trainer/dpo_trainer.py", line 635, in __init__
    self.tokenizer = tokenizer
    ^^^^^^^^^^^^^^
AttributeError: property 'tokenizer' of 'DPOTrainer' object has no setter
```

### Expected behavior

To work, like before.

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes #2161 

Ensure backward compatibility for DPO and SFT only

```python
>>> from datasets import load_dataset
>>> from transformers import AutoModelForCausalLM, AutoTokenizer
>>> from trl import DPOConfig, DPOTrainer
>>> model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
>>> tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
>>> dataset = load_dataset("trl-lib/Capybara-Preferences", split="train")
>>> training_args = DPOConfig(output_dir="Qwen2.5-0.5B-DPO")
>>> trainer = DPOTrainer(model=model, args=training_args, train_dataset=dataset, tokenizer=tokenizer)
/fsx/qgallouedec/miniconda3/envs/trl/lib/python3.11/site-packages/huggingface_hub/utils/_deprecation.py:101: FutureWarning: `tokenizer` is deprecated and will be removed in version 0.14.0 for `DPOTrainer.__init__`. Use `processing_class` instead.
  return f(*args, **kwargs)
```

## TODO

- [x] BCO
- [x] CPO
- [x] DPO
- [x] GKD
- [x] KTO
- [x] Nash-MD
- [x] Online DPO
- [x] ORPO
- [x] PPOv2
- [x] Reward
- [x] RLOO
- [x] SFT
- [x] XPO

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -670,7 +680,7 @@ def make_inputs_require_grad(module, input, output):
         else:
             self.ref_model = create_reference_model(model)
 
-        if tokenizer is None:
+        if processing_class is None:
             raise ValueError("tokenizer must be specified to tokenize a DPO dataset.")
```
Comment (alvarobartt) [inline trl/trainer/dpo_trainer.py:684]:
```suggestion
            raise ValueError("processing_class must be specified to tokenize a DPO dataset.")
```

File: trl/trainer/ppov2_trainer.py [hunk 2]
```diff
@@ -215,7 +220,7 @@ def __init__(
             self.train_dataset,
             batch_size=self.local_dataloader_batch_size,
             shuffle=True,
-            collate_fn=DataCollatorWithPadding(tokenizer),
+            collate_fn=DataCollatorWithPadding(processing_class),
```
Comment (alvarobartt) [inline trl/trainer/ppov2_trainer.py:223]:
```suggestion
            collate_fn=DataCollatorWithPadding(self.processing_class),
```

File: trl/trainer/ppov2_trainer.py [hunk 3]
```diff
@@ -599,7 +604,7 @@ def repeat_generator():
 
     def generate_completions(self, sampling: bool = False):
         args = self.args
-        tokenizer = self.tokenizer
+        processing_class = self.processing_class
```
Comment (alvarobartt) [inline trl/trainer/ppov2_trainer.py:607]: I may be missing something but is this required? Cannot we just use `self.processing_class`?

File: trl/trainer/rloo_trainer.py [hunk 4]
```diff
@@ -144,7 +149,7 @@ def __init__(
         for module in [policy, ref_policy, reward_model]:
             disable_dropout_in_model(module)
         if args.stop_token and args.stop_token == "eos":
-            args.stop_token_id = tokenizer.eos_token_id
+            args.stop_token_id = processing_class.eos_token_id
```
Comment (alvarobartt) [inline trl/trainer/rloo_trainer.py:152]:
```suggestion
            args.stop_token_id = self.processing_class.eos_token_id
```

File: trl/trainer/rloo_trainer.py [hunk 5]
```diff
@@ -187,7 +192,7 @@ def __init__(
             self.train_dataset,
             batch_size=self.local_dataloader_batch_size,
             shuffle=True,
-            collate_fn=DataCollatorWithPadding(tokenizer),
+            collate_fn=DataCollatorWithPadding(processing_class),
```
Comment (alvarobartt) [inline trl/trainer/rloo_trainer.py:195]:
```suggestion
            collate_fn=DataCollatorWithPadding(self.processing_class),
```

File: trl/trainer/ppov2_trainer.py (hunk 3 above)
Comment (qgallouedec) [reply trl/trainer/ppov2_trainer.py:607]: You're right. Same for `args`. It will probably need some refactoring in the future

Comment (alvarobartt) [reply trl/trainer/ppov2_trainer.py:607]: I've seen that in other places too so maybe there's a rationale that I don't see for that? Not sure, but sure we'll keep it in mind

**Review by edbeeching — APPROVED** [review]

**Review by alvarobartt — APPROVED** [review]

**Comment by yananchen1989** [comment]
> Traceback (most recent call last):
>   File "/workspace/trl/examples/scripts/sft.py", line 93, in <module>
>     trainer = SFTTrainer(
>   File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_deprecation.py", line 101, in inner_f
>     return f(*args, **kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/transformers/utils/deprecation.py", line 165, in wrapped_func
>     return func(*args, **kwargs)
>   File "/workspace/trl/trl/trainer/sft_trainer.py", line 409, in __init__
>     super().__init__(
> TypeError: Trainer.__init__() got an unexpected keyword argument 'processing_class'

**Comment by yananchen1989** [comment]
trl version: 0.12.0.dev0

**Comment by kashif** [comment]
you need to use the main version of `transformers`

**Comment by BUILDERlym** [comment]
> you need to use the main version of `transformers`

got it, thanks

**Pull request merged**